kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt

```diff
@@ -5,128 +5,498 @@
 
 package space.kscience.kmath.functions
 
-import space.kscience.kmath.misc.UnstableKMathAPI
-import space.kscience.kmath.operations.*
-import kotlin.contracts.InvocationKind
-import kotlin.contracts.contract
-import kotlin.math.max
-import kotlin.math.pow
+import space.kscience.kmath.operations.Ring
+import space.kscience.kmath.operations.invoke
+import kotlin.js.JsName
+import kotlin.jvm.JvmName
 
-/**
- * Polynomial coefficients model without fixation on specific context they are applied to.
- *
- * @param coefficients constant is the leftmost coefficient.
- */
-public class Polynomial<out T>(public val coefficients: List<T>) {
-    override fun toString(): String = "Polynomial$coefficients"
-}
 
 /**
- * Returns a [Polynomial] instance with given [coefficients].
+ * Abstraction of polynomials.
  */
-@Suppress("FunctionName")
-public fun <T> Polynomial(vararg coefficients: T): Polynomial<T> = Polynomial(coefficients.toList())
+public interface Polynomial<C>
 
 /**
- * Evaluates the value of the given double polynomial for given double argument.
+ * Abstraction of ring of polynomials of type [P] over ring of constants of type [C].
+ *
+ * @param C the type of constants. Polynomials have them as coefficients in their terms.
+ * @param P the type of polynomials.
  */
-public fun Polynomial<Double>.value(arg: Double): Double = coefficients.reduceIndexed { index, acc, c ->
-    acc + c * arg.pow(index)
-}
+@Suppress("INAPPLICABLE_JVM_NAME", "PARAMETER_NAME_CHANGED_ON_OVERRIDE") // FIXME: Waiting for KT-31420
+public interface PolynomialSpace<C, P: Polynomial<C>> : Ring<P> {
```

Comment:
This class seems to be a needless abstraction. I do not see, where all of its functions are used.

Reply:
For me historically it was a systematisation of API that I want to see in polynomial spaces. I also thought there could be cases of need in common abstraction (for some extension functions or something other). But currently there is no actual use case except for systematisation of the API.